File: packages/cli/src/commands/models-dev.ts
import { Command } from 'commander'
import { withGracefulShutdown, createPathMap, type PathMap } from '../common.ts'
import path from 'node:path'
import fs from 'node:fs/promises'

export const modelsDev = new Command('models-dev')

modelsDev.command('providers').action(() =>
  withGracefulShutdown(async signal => {
    const pathMap = await createPathMap()
    const data = await fetchData({ signal, pathMap })
    const providers = Object.keys(data).toSorted()
    console.log(JSON.stringify(providers, null, 2))
  }),
)

modelsDev
  .command('models')
  .argument('<provider>')
  .action(provider =>
    withGracefulShutdown(async signal => {
      const pathMap = await createPathMap()
      const data = await fetchData({ signal, pathMap })
      const providerData = data[provider] ?? {}
      const models = Object.keys(providerData.models ?? {}).toSorted()
      console.log(JSON.stringify(models, null, 2))
    }),
  )

modelsDev
  .command('model')
  .argument('<provider>')
  .argument('<model>')
  .action((provider, model) =>
    withGracefulShutdown(async signal => {
      const pathMap = await createPathMap()
      const data = await fetchData({ signal, pathMap })
      const providerData = data[provider] ?? {}
      const modelData = providerData.models?.[model] ?? {}
      console.log(JSON.stringify(modelData, null, 2))
    }),
  )

type Metadata = { etag?: string }

async function fetchData(input: {
  signal: AbortSignal
  pathMap: PathMap
  attempt?: number
  maxAttempts?: number
}): Promise<Record<string, any>> {
  const { signal, pathMap, attempt = 1, maxAttempts = 3 } = input

  if (attempt > maxAttempts) {
    throw new Error(`Failed to fetch models after ${maxAttempts} attempts.`)
  }

  const { repoRoot } = pathMap
  const configDir = path.join(repoRoot, '.nailed-it')
  const metadataPath = path.join(configDir, 'models-dev-metadata.json')
  const dataPath = path.join(configDir, 'models-dev-data.json')

  let metadata: Metadata = {}
  try {
    const content = await fs.readFile(metadataPath, 'utf8')
    metadata = JSON.parse(content) as Metadata
  } catch {
    await fs.rm(metadataPath, { force: true })
  }

  let headers: Record<string, string> = {}
  if (metadata.etag) {
    headers['If-None-Match'] = metadata.etag
  }

  const response = await fetch('https://models.dev/api.json', {
    signal,
    headers,
  })

  if (response.status === 304) {
    let data: Record<string, any> = {}
    try {
      const content = await fs.readFile(dataPath, 'utf8')
      data = JSON.parse(content)
    } catch {
      await fs.rm(dataPath, { force: true })
      await fs.rm(metadataPath, { force: true })
      return await fetchData({
        signal,
        pathMap,
        attempt: attempt + 1,
        maxAttempts,
      })
    }

    console.error('Models are up to date.')
    return data
  }

  if (!response.ok) {
    const text = await response.text()
    throw new Error(
      `Failed to fetch models: ${response.status} ${response.statusText} - ${text}`,
    )
  }

  let newMetadata: Metadata = {}
  const etag = response.headers.get('etag')
  if (etag) {
    newMetadata.etag = etag
  }

  const data = await response.json()

  await fs.mkdir(configDir, { recursive: true })
  await fs.writeFile(dataPath, JSON.stringify(data, null, 2))
  await fs.writeFile(metadataPath, JSON.stringify(newMetadata, null, 2))

  console.error('Models updated successfully.')
  return data
}

File: python-packages/dev-cli/src/dev_cli/commands/models_dev.py
import json
import sys
import urllib.error
import urllib.request
from typing import Any

import typer
from dev_cli.common import REPO_ROOT, check_repo_root

app = typer.Typer(no_args_is_help=True)

CONFIG_DIR = REPO_ROOT / ".nailed-it"
METADATA_PATH = CONFIG_DIR / "models-dev-metadata.json"
DATA_PATH = CONFIG_DIR / "models-dev-data.json"
API_URL = "https://models.dev/api.json"
_HTTP_NOT_MODIFIED = 304


def _fetch_data(
    *,
    attempt: int = 1,
    max_attempts: int = 3,
) -> dict[str, Any]:
    if attempt > max_attempts:
        msg = f"Failed to fetch models after {max_attempts} attempts."
        raise RuntimeError(msg)

    check_repo_root()

    metadata: dict[str, str] = {}
    try:
        metadata = json.loads(METADATA_PATH.read_text("utf-8"))
    except FileNotFoundError, json.JSONDecodeError:
        METADATA_PATH.unlink(missing_ok=True)

    headers: dict[str, str] = {
        "User-Agent": "dev-cli/0.0.0",
    }
    if etag := metadata.get("etag"):
        headers["If-None-Match"] = etag

    req = urllib.request.Request(API_URL, headers=headers)

    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            data: dict[str, Any] = json.loads(resp.read())

            new_metadata: dict[str, str] = {}
            if etag := resp.headers.get("ETag"):
                new_metadata["etag"] = etag

            CONFIG_DIR.mkdir(parents=True, exist_ok=True)
            DATA_PATH.write_text(json.dumps(data, indent=2, ensure_ascii=False))
            METADATA_PATH.write_text(
                json.dumps(new_metadata, indent=2, ensure_ascii=False),
            )

            print("Models updated successfully.", file=sys.stderr)
            return data

    except urllib.error.HTTPError as e:
        if e.code == _HTTP_NOT_MODIFIED:
            try:
                data = json.loads(DATA_PATH.read_text("utf-8"))
            except FileNotFoundError, json.JSONDecodeError:
                DATA_PATH.unlink(missing_ok=True)
                METADATA_PATH.unlink(missing_ok=True)
                return _fetch_data(
                    attempt=attempt + 1,
                    max_attempts=max_attempts,
                )
            else:
                print("Models are up to date.", file=sys.stderr)
                return data

        text = e.read().decode("utf-8")
        msg = f"Failed to fetch models: {e.code} {e.reason} - {text}"
        raise RuntimeError(msg) from e


@app.command()
def providers() -> None:
    """List all provider names."""
    data = _fetch_data()
    providers_list = sorted(data)
    print(json.dumps(providers_list, indent=2, ensure_ascii=False))


@app.command()
def models(provider: str) -> None:
    """List model names for a provider."""
    data = _fetch_data()
    provider_data: dict[str, Any] = data.get(provider, {})
    models_list = sorted(provider_data.get("models", {}))
    print(json.dumps(models_list, indent=2, ensure_ascii=False))


@app.command()
def model(provider: str, model: str) -> None:
    """Show model details."""
    data = _fetch_data()
    provider_data: dict[str, Any] = data.get(provider, {})
    model_data: dict[str, Any] = provider_data.get("models", {}).get(model, {})
    print(json.dumps(model_data, indent=2, ensure_ascii=False))
